### PR TITLE
rewrite range map

### PR DESCRIPTION
**Pull request by RalfJung**
This significantly improves the benchmark in the OP of https://github.com/solson/miri/issues/593. (I wonder if we should stick that somewhere, also use it in the future as a stress test for the memory subsystem, or whatever else it is stressing?)

Runtime improves by almost 40% (39s -> 23s), memory usage improves by over 90% (2700MB -> 250 MB).  I measured this using `/usr/bin/time -v cargo +nightly miri run`, looking for the "User time" and "Maximum resident set size".

Fixes https://github.com/solson/miri/issues/593.

### CODE REVIEW

File: src/range_map.rs [hunk 1]
```diff
@@ -6,137 +6,112 @@
...
+        // Copy the data, and insert 2nd element
+        let second = Elem {
+            range: second_range,
+            data: elem.data.clone(),
+        };
+        self.v.insert(index+1, second);
+        return true;
     }
 
     /// Provide mutable iteration over everything in the given range.  As a side-effect,
```
Comment (oli-obk) [inline src/range_map.rs]: a further side effect not mentioned in the docs is that adjoining elements are merged

**Comment by oli-obk** [comment]
r=me with docs nit fixed

**Comment by RalfJung** [comment]
According to https://github.com/solson/miri/issues/593#issuecomment-451740772 this is a perf regression for another benchmark, so I'll need to do more investigation.

Not sure when I will find the time.

**Comment by RalfJung** [comment]
I could reproduce that regression locally (not 40x but still 18x or so). It's real.

Not sure when I'll have time to investigate, though.

**Comment by RalfJung** [comment]
I updated this PR so that both benchmarks look good now.  There's a heuristic in range_map that stops merging after too many unsuccessful attempts.

**Pull request merged**